power_grid_model_c/power_grid_model/include/power_grid_model/sparse_mapping.hpp

```diff
@@ -50,35 +50,164 @@ struct SparseMapping {
         return SparseEntry{i_A, j_B};
     });
     // initialize result
     SparseMapping sparse_mapping;
+    // size of indpntr will become 8
     sparse_mapping.indptr.resize(n_B + 1);
+    // size of reorder will be 6
     sparse_mapping.reorder.resize(n_A);
     // counting sort column, keep row as sorted
-    IdxVector counter(n_B, 0);
+    // counter = {n_Bx0} = {0, 0, 0, 0, 0, 0, 0}
+    IdxVector counter(n_B, 0); // amount of load gens connected to each bus
+    // For every entry, ex. {0,3}, we add 1 to counter at entry.second place in this case 3rd, so it becomes {0, 0, 0,
+    // 1, 0, 0, 0} counter should become {0, 2, 2, 1, 0, 1, 0}
     for (auto const& entry : entries) {
         ++counter[entry.second];
     }
     // accumulate value to indptr
+    // [a, b, c, d], the inclusive scan would produce [a, a+b, a+b+c, a+b+c+d]
+    // counter = {0, 2, 2, 1, 0, 1, 0}, indptr becomes {0, 0, 2, 4, 5, 5, 6, 6}
     std::inclusive_scan(counter.cbegin(), counter.cend(), sparse_mapping.indptr.begin() + 1);
     // copy back
+    // copy everything except the indptr[0] to counter.
+    // counter becomes {0, 2, 4, 5, 5, 6, 6}
     std::copy(sparse_mapping.indptr.cbegin() + 1, sparse_mapping.indptr.cend(), counter.begin());
     // sort
+    // crbegin() points to the element at the end of the container
+    // crend() points to the position before the first element
+    // at this point: entries[{0,3},{1,5},{2,2},{3,1},{4,1},{5,2}], counter {0, 2, 4, 5, 5, 6, 6}
+    // Reverse order the first entry in entries: {5,2}, it_entry->second = 2
+    // counter[2] is 4, decremented to 3(counter becomes {0, 2, 3, 5, 5, 6, 6}), thus sparse_mapping.reorder[3] is set
+    // to it_entry->first which is 5. sparse_mapping.reorder = {0, 0, 0, 5, 0, 0},
+    // Final result: sparse_mapping.reorder = {3, 4, 2, 5, 0, 1}
     for (auto it_entry = entries.crbegin(); it_entry != entries.crend(); ++it_entry) {
         sparse_mapping.reorder[--counter[it_entry->second]] = it_entry->first;
     }
+    // the first is the index before reordering of load_gen
+    // it entry second is the index of the bus
+    // reorder at the end contains for each new index of the load gen the old index of the load gen in a linear way
     assert(sparse_mapping.indptr[0] == 0);
     assert(sparse_mapping.indptr.back() == n_A);
     return sparse_mapping;
 }
 
-} // namespace power_grid_model
+struct DenseMapping {
+    IdxVector indvector;
+    IdxVector reorder;
+};
+
+inline DenseMapping build_dense_mapping(IdxVector const& idx_B_in_A, Idx const n_B) {
+    auto const n_A = static_cast<Idx>(idx_B_in_A.size());
+    auto const r = std::max_element(idx_B_in_A.begin(), idx_B_in_A.end())[0];
+    // for counting sort nr_of_operations = 1 + 3*n_A + 2*n_B + 2*n_A + 2*n_B - 1 + 1 + n_A + 1 + 14 * n_A
+    // nr_of_operations = 20*n_A + 4*n_B + 2
+    auto const c = 20 * n_A + 4 * n_B + 2;
+    // for n log n sort nr of operations = 1 + 7*n_A + n_A*log(n_A)?
```

Comment:
this seems reasonable. it's nice to see the approximate form of the formula being so simple as they are.

based on this, the criterion reduces to something approximately like `n_B < ((prefactor/4) * log(n_A) - 13/4) n_A`, or (approximating and simplifying even further) `n_B / n_A < c * log(n_A) - 3`

the important thing now is to measure the prefactor (or, equivalently, `c`) before `n_A log n_A`.

Because we can assume that, usually, either `n_A = 0, 1, 2, 3` (**MUCH** smaller than `n_B`), or `n_A` is of the order of `n_B / 10` or `n_B / 100`, or otherwise otherwise `n_A ~ n_B` or `n_A ~ 10 n_B`

* In the first case, then `n_B / n_A` will be large and the prefactor doesn't matter.
* In the other cases, the prefactor is important, because the threshold depends on it. If it is large, basically for any `n_A` for which `n_B / n_A` is small, the counting sort will be faster, but if it is small, that will only be true for large `n_A`